bitcoin/BRSyncManager.c

```diff
@@ -1162,25 +1164,23 @@ BRClientSyncManagerAnnounceGetTransactionsItem (BRClientSyncManager manager,
     }
 
     if (needRegistration) {
-        if (NULL != BRWalletTransactionForHash (manager->wallet, transaction->txHash)) {
-            // Wallet already knows about this txn; so just update the block info
-            BRWalletUpdateTransactions (manager->wallet, &transaction->txHash, 1, (uint32_t) blockHeight, (uint32_t) timestamp);
-        } else {
-            // Set the transaction's block info according to what has been announced
-            transaction->timestamp = (uint32_t) timestamp;
-            transaction->blockHeight = (uint32_t) blockHeight;
-
+        if (NULL == BRWalletTransactionForHash (manager->wallet, transaction->txHash)) {
             // BRWalletRegisterTransaction doesn't reliably report if the txn was added to the wallet.
             BRWalletRegisterTransaction (manager->wallet, transaction);
-            if (BRWalletTransactionForHash (manager->wallet, transaction->txHash) == transaction) {
+            if (transaction == BRWalletTransactionForHash (manager->wallet, transaction->txHash)) {
                 // If our transaction made it into the wallet, do not deallocate it
-                transaction = NULL;
+                needFree = 0;
             }
         }
     }
 
+    // Wallet already knows about this txn; so just update the block info.
+    if (/* isRegistered */ NULL != transaction && BRTransactionIsSigned (transaction)) {
+        BRWalletUpdateTransactions (manager->wallet, &transaction->txHash, 1, (uint32_t) blockHeight, (uint32_t) timestamp);
```

Comment:
This should be in the above `needRegistration` if block, ya? (obviates the need to repeat the NULL and signed check, also respects the request ID being correct.

Reply:
Don't we want to update the transaction even if we don't need to register it?

Reply:
I updated the logic with a clear check for if the update should occur.